zebrad/src/components/mempool/downloads.rs

```diff
@@ -0,0 +1,284 @@
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use color_eyre::eyre::eyre;
+use futures::{
+    future::TryFutureExt,
+    ready,
+    stream::{FuturesUnordered, Stream},
+};
+use pin_project::pin_project;
+use tokio::{sync::oneshot, task::JoinHandle};
+use tower::{Service, ServiceExt};
+use tracing_futures::Instrument;
+
+use zebra_chain::transaction::UnminedTxId;
+use zebra_consensus::transaction as tx;
+use zebra_network as zn;
+use zebra_state as zs;
+
+type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+/// The maximum number of concurrent inbound download and verify tasks.
+///
+/// We expect the syncer to download and verify checkpoints, so this bound
+/// can be small.
+///
+/// ## Security
+///
+/// We use a small concurrency limit, to prevent memory denial-of-service
+/// attacks.
+///
+/// The maximum transaction size is 2 million bytes. A deserialized malicious
+/// transaction with ~225_000 transparent outputs can take up 9MB of RAM. As of
+/// February 2021, a growing `Vec` can allocate up to 2x its current length,
+/// leading to an overall memory usage of 18MB per malicious transaction. (See
+/// #1880 for more details.)
```

Comment:
I think we fixed the vector growth issue using a preallocation size trait.

Can you fix both these docs and the syncer docs?

```suggestion
/// The maximum transaction size is 2 million bytes. A deserialized malicious
/// transaction with ~225_000 transparent outputs can take up 9MB of RAM.
/// (See #1880 for more details.)
```